### ISSUE TITLE
R LSP crash

### ISSUE THREAD

**Issue by jmcphers** [posit-dev/positron#5507]
Repro unknown, but we have observed this R LSP crash on Workbench (added linebreaks for readability):

```
[R]   2024-11-25T19:46:24.079689Z ERROR
Panic! In file 'crates/ark/src/lsp/document_context.rs' at line 32:
Failed to find closest node to point: (0, 45)
with contents 'mtcars |> ggplot(aes(x = disp,y=mpg)) + geom'
[R] 
[R] Backtrace:
[R]    0: <unknown>
[R]    1: <unknown>
[R]    2: <unknown>
[R]    3: <unknown>
[R]    4: <unknown>
[R]    5: <unknown>
```

@jthomasmock observed this with a new R session, at which he typed `library(tidyverse)`, and after the tidyverse loaded, typed `mtcars |> ggplot(aes(x = disp,y=mpg)) + geom`.

It appeared that the LSP crashed when trying to generate suggestions for symbols completing the text `geom`. 

Even if we can't reproduce it, it seems possible to make the LSP handle this error gracefully instead of crashing. 

Ark version info from this environment (PTD, ark 0.1.153):

```
 branch 
                                                                               "main" 
                                                                               commit 
                                                                            "6e40dec" 
                                                                                 date 
                                                            "2024-11-14 20:13:02 UTC" 
                                                                               flavor 
                                                                            "release" 
                                                                                 path 
"/usr/lib/rstudio-server/bin/positron-server/extensions/positron-r/resources/ark/ark" 
                                                                              version 
                                                                            "0.1.153" 
```

**Comment by lionel-** [comment]
I agree this should only be a crash in dev builds.

For resiliency I wonder if we should add a way for the server to ask the client to restart the LSP when we detect an inconsistency in our state. This would prevent cascading issues arising from the inconsistency.

It would be useful to record these occurrences in telemetry.

**Comment by lionel-** [comment]
That's the first report about an inconsistent state (position OOB by 1 character, suggesting we missed an update notification before receiving the completion request) since we have refactored the LSP. Even though we removed all stopgap measures for reordering notifications coming out of order. So I wonder if something is specifically wrong on Workbench rather than in the LSP.

@jmcphers Do you see any obvious reason that could explain out of order or missing LSP messages on Workbench?

**Comment by jmcphers** [comment]
No, there's nothing about Workbench that I can think of that would affect this. The communication between the LSP and the extension host is over TCP and it's on the same host as ark, so it should just be the same local network stack you'd get in dev mode (i.e. workbench is not in the loop). 

We _did_ see a bunch of extension host issues on Workbench; I don't see them as likely to be related to this issue, but it isn't impossible. See #5508 for more details.

@lionel- for now, would you be convert this to a warning (with appropriate logging) instead of a crash?

**Comment by DavisVaughan** [comment]
These crashes in particular typically communicate that we have something wrong in our LSP model, as otherwise IIUC the LSP should never allow documents to get out of sync, so keeping them crashes has helped us resolve 3-4 critical issues on the ark side of the LSP that otherwise we never would have heard about.

I also think that converting to a warning isn't particularly helpful because this isn't really a recoverable error for that file, once its out of sync, its going to be impossible to sync it back up with what the client thinks the state of the file is.

So there's probably something else going on under here that we really would like to figure out

**Comment by lionel-** [comment]
@jmcphers I think it should be more than a warning because the LSP is in an inconsistent state at this point and there is no way to recover.

Here is how I propose to address such inconsistency issues:

- Record information about the problematic state and disable the LSP (ignore any further messages). The kernel lives on.
- Send a crash notification to Positron's UI comm. This notification is surfaced to the user with a request to file an issue with the debug info.
- At the same time, Positron restarts the LSP. Only the client can do that so Positron is the initiator. This will require backporting the tower-lsp fixes to the shutdown procedure we implemented in air.
- It we get a crash notification again, we let the user know the LSP crashed a second time (more info in the output channel) and will not be restarted.

This is similar to how the vscode client for LSP restarts the server up to three times. Except it's completely silent and doesn't notify users. We would be intentionally more verbose in order to encourage error reporting.

**Comment by jmcphers** [comment]
I just hit this again, after typing in "23" and executing it in the R console (again on Palm PTD). 

```
R]   2025-01-30T16:14:04.976783Z ERROR  Panic! In file 'crates/ark/src/lsp/document_context.rs' at line 32: Failed to find closest node to point: (0, 5) with contents ''
[R] 
```

With this ark version:

```
> .ps.ark.version()
                                                                               branch 
                                                                               "main" 
                                                                               commit 
                                                                            "a569d6e" 
                                                                                 date 
                                                            "2024-12-18 12:29:27 UTC" 
                                                                               flavor 
                                                                            "release" 
                                                                                 path 
"/usr/lib/rstudio-server/bin/positron-server/extensions/positron-r/resources/ark/ark" 
                                                                              version 
                                                                            "0.1.159" 
```

**Comment by lionel-** [comment]
With https://github.com/posit-dev/ark/pull/679 and https://github.com/posit-dev/positron/pull/6181, Ark no longer crashes when the LSP panics. Instead it is shut down and a toast notification is sent to the client to let the user know LSP features are disabled and we'd love a bug report with logs.

The particular crash discussed here has not been observed in desktop sessions AFAIK. We'll need a deeper look at interactions between Workbench, Positron, and Ark / language servers and why messages could be missed or arrive out of order in that setup.

**Comment by juliasilge** [comment]
We believe this is no longer a problem.
